Outbound Foreign Keys

RFSH opened the pull request:
This PR adds the support for outbound foreign keys in reference.

Main changes are described in #203 (first comment). These are the lines that you can find them:
 - part 1 (#171): [changing the read call uri](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/reference.js#L563,L600), [attaching FKR values to page](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/reference.js#L1312,L1323)
 - part 2 ([Chaise Issue #654](https://github.com/informatics-isi-edu/chaise/issues/654)): [new contextualize](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/core.js#L1540,L1640), [PseudoColumn](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/core.js#L1894,L2015), [Reference.columns](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/reference.js#L263,L268)
 - part 3 (#159): [Tuple.values](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/reference.js#L1708,L1716), [PseudoColumn formatPresentation](https://github.com/informatics-isi-edu/ermrestjs/blob/attribute-group/js/core.js#L1994,L2010)

These are the other changes that were needed to be made in order to complete the task:
- Change the Reference constructor and `Reference.resolve` to be able to create Reference synchronously (#229).
- Add `InboundForeignKeys` class to represent inbound foreign keys and relocate the `_contextualize` to this class.
- Create `_generateRowName` function in utilities. Since it's being used in multiple places.

shinyichen commented on js/core.js:
```diff
@@ -1572,34 +1536,105 @@ var ERMrest = (function (module) {
-                        columns._push(c);
+                        visiblePseudoColumns.push(col);
+                        
+                    } catch (exception) {}
```
Empty exception handler?

shinyichen commented on js/core.js:
```diff
@@ -1853,6 +1890,129 @@ var ERMrest = (function (module) {
+         * @type {boolean}
+         * @desc indicator that this is a PseudoColumn rather than a Column.
+         */
+        this.isPseudo = true;
```
Add the flag to Column as well.

shinyichen commented on js/reference.js:
```diff
@@ -870,6 +947,7 @@ var ERMrest = (function(module) {
                     newRef = _referenceCopy(this);
                     delete newRef._context; // NOTE: related reference is not contextualized
                     delete newRef._related;
+                    delete newRef._pseudoColumns;
```
Should this line be moved to `.setNewTable()` method?

shinyichen replied:
Also add this line to `.setNewTable()` method?

RFSH commented:
I addressed the comments that @shinyichen and @jrchudy made. It's ready to merge.

Merging this will cause recordset to throw an error when user sorts using a `PseudoColumn`.  Therefore we should first merge [this PR in chaise](https://github.com/informatics-isi-edu/chaise/pull/686). 

However we should address it later and think about a better solution rather than just disabling (#246).

shinyichen commented on js/reference.js:
```diff
@@ -544,6 +564,62 @@ var ERMrest = (function(module) {
+
+                    // add M alias to current table
+                    linking = parts[parts.length-1].match(/(\(.*\)=\(.*:.*:.*\))/);
+                    if (linking && linking[1]) { // the same logic as parser for finding the link syntax
```
Just note to myself and @RFSH: Update this line to take search filter into account. See latest parser code.
